llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp

```diff
@@ -3109,6 +3109,32 @@ Value *LibCallSimplifier::optimizeRemquo(CallInst *CI, IRBuilderBase &B) {
   return ConstantFP::get(CI->getType(), Rem);
 }
 
+/// Constant folds fdim
+Value *LibCallSimplifier::optimizeFdim(CallInst *CI, IRBuilderBase &B) {
+  const APFloat *X, *Y;
+  // Check if both values are constants
+  if (!match(CI->getArgOperand(0), m_APFloat(X)) ||
+      !match(CI->getArgOperand(1), m_APFloat(Y)))
+    return nullptr;
+  // If either argument is NaN, NaN is returned
+  if (X->isNaN() || Y->isNaN())
+    return ConstantFP::getQNaN(CI->getType());
+
+  IRBuilderBase::FastMathFlagGuard Guard(B);
+  FastMathFlags FMF = CI->getFastMathFlags();
+  // set no-NaN fast-math-flag as we already checked for NaN for both operands
+  FMF.setNoNaNs();
+  // set no-signed-zeroes as fdim will never return -0.0
+  FMF.setNoSignedZeros();
+  B.setFastMathFlags(FMF);
+  // fdim is equivalent to fmax(x - y, 0), except for the NaN handling requirements. 
```

Comment:
If I read this correctly, this wants to propagate nans instead of the return non-nan operand behavior. Can this just unconditionally use Intrinsic::maximum instead?

Reply:
this was explicitly mentioned in cppref notes https://en.cppreference.com/w/cpp/numeric/math/fdim
so i assumed maximum intrinsic does not handles this

documentation for maxnum says the same
https://llvm.org/docs/LangRef.html#llvm-maxnum-intrinsic
"If either operand is a NaN, returns the other non-NaN operand. Returns NaN only if both operands are NaN. If the operands compare equal, returns either one of the operands. For example, this means that fmax(+0.0, -0.0) returns either -0.0 or 0.0."

Reply:
Maximum, not maxnum. https://llvm.org/docs/LangRef.html#llvm-maximum-intrinsic, which is also fmaximum in C23

Reply:
maximum also has stricter signed 0 behavior than maxnum

Reply:
oh, i misread maxnum
yeah, maximum intrinsic actually fits the box here, thanks for the pointer

Reply:
I would suggest trying to use maxnum if the call is nnan, and if not just use maximum

Reply:
Actually maybe just use maximum, the nnan case can transform later if needed. I'm more wondering if it is OK to add nsz to specify the relaxed zero requirement